### DISCUSSION TITLE
maximum recursion depth exceeded while calling a Python object

### DISCUSSION THREAD

**Discussion by alpxe** [pytransitions/transitions#613] [Q&A]
from transitions import State, Machine

class Matter:

    def helloAAA(self):
        print(f"hello {self.state}")
        self.toB()

    def helloBBB(self):
        print(f"hello {self.state}")
        self.toC()

    def helloCCC(self):
        print(f"hello {self.state}")
        self.toA()

    def __init__(self):
        states = [
            State(name='S', on_enter='say_hello'),
            State(name='A', on_enter='helloAAA'),
            State(name='B', on_enter='helloBBB'),
            State(name='C', on_enter='helloCCC'),
        ]
        self.machine = Machine(model=self, states=states, initial="S")

        self.machine.add_transition(trigger='starting', source='S', dest='A')
        self.machine.add_transition(trigger='toB', source='A', dest='B')
        self.machine.add_transition(trigger='toC', source='B', dest='C')
        self.machine.add_transition(trigger='toA', source='C', dest='A')

lump = Matter()

lump.starting()

Unable to loop??? 
The requirement are step by step , then exit
but it is wrong!!!!!

**Comment by alpxe** [comment]
lump.starting()

while True:
    if lump.state == "A":
        lump.toB()
    elif lump.state == "B":
        lump.toC()
    elif lump.state == "C":
        lump.toA()

and  >>>  This is right

**Comment by aleneum** [comment]
Hello @alpxe,

your example above translates to:

```python
def enterA():
    enterB()

def enterB():
    enterC()

def enterC():
    enterA()
```

while the code in your second example does this:

```
enterB()
enterC()
enterA()
```

The `enter` callbacks in your first case basically never return which fills the stack quite quickly. You can, however, trigger events inside callbacks when you instruct transitions to process machine events sequentially instead of instantly by passing `queued=True` to the Machine constructor

```python
from transitions import Machine, State

class Foo:
    def helloAAA(self):
        print(f"hello {self.state}")
        self.toB()

    def helloBBB(self):
        print(f"hello {self.state}")
        self.toC()

    def helloCCC(self):
        print(f"hello {self.state}")
        self.toA()

    def __init__(self):
        states = [
            State(name='S', on_enter='say_hello'),
            State(name='A', on_enter='helloAAA'),
            State(name='B', on_enter='helloBBB'),
            State(name='C', on_enter='helloCCC'),
        ]
        self.machine = Machine(model=self, states=states, initial="S", queued=True)   # <--- queued=True

        self.machine.add_transition(trigger='starting', source='S', dest='A')
        self.machine.add_transition(trigger='toB', source='A', dest='B')
        self.machine.add_transition(trigger='toC', source='B', dest='C')
        self.machine.add_transition(trigger='toA', source='C', dest='A')

f = Foo()
f.starting()
```
